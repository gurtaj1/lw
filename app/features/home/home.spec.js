import React from 'react';
import { shallow } from 'enzyme';

import Home, { negativeBalanceCheck } from 'features/home/home';

describe('features.home', () => {
  describe('Home', () => {
    it('should export something', () => {
      expect(Home).toBeDefined();
    });

    it('should render without crashing', () => {
      const wrapper = shallow(<Home />);

      expect(wrapper).toMatchSnapshot();
    });

    describe('negativeBalanceCheck()', () => {
      window.alert = jest.fn();

      afterEach(() => {
        jest.clearAllMocks();
      });

      it('should give warning when entering negative balance', () => {
        negativeBalanceCheck(0, -1);

        expect(window.alert).toHaveBeenCalled();
      });

      it('should not give warning when remaining in negative balance', () => {
        negativeBalanceCheck(-1, -2);

        expect(window.alert).not.toHaveBeenCalled();
      });

      it('should not give warning when entering non-negative balance', () => {
        negativeBalanceCheck(-1, 0);

        expect(window.alert).not.toHaveBeenCalled();        
      });

      it('should not give warning when entering positive balance', () => {
        negativeBalanceCheck(0, 1);

        expect(window.alert).not.toHaveBeenCalled();        
      });

      it('should not give warning when remaining in positive balance', () => {
        negativeBalanceCheck(1, 2);

        expect(window.alert).not.toHaveBeenCalled();        
      });
    });

    describe('balance button click functions', () => {
      it('should call decreaseBalance when the minus button is clicked', () => {
        const spiedFunction = jest.spyOn(Home.prototype, 'decreaseBalance');
        const wrapper = shallow(<Home />);
        wrapper.find('.balance-buttons').childAt(0).simulate('click');

        expect(spiedFunction).toHaveBeenCalled();
      });

      it('should decrease state balance when decreaseBalance() is called', () => {
        const wrapper = shallow(<Home />);
        wrapper.find('.balance-buttons').childAt(0).simulate('click');

        expect(wrapper.state('balance')).toBe(11);
      });

      it('should call increaseBalance when the plus button is clicked', () => {
        const spiedFunction = jest.spyOn(Home.prototype, 'increaseBalance');
        const wrapper = shallow(<Home />);
        wrapper.find('.balance-buttons').childAt(1).simulate('click');

        expect(spiedFunction).toHaveBeenCalled();
      });

      it('should increase state balance when increaseBalance() is called', () => {
        const wrapper = shallow(<Home />);
        wrapper.find('.balance-buttons').childAt(1).simulate('click');

        expect(wrapper.state('balance')).toBe(13);
      });
    });
  });
});
