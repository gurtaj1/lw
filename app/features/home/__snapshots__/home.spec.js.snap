// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features.home Home should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "Your Balance: ",
        "£12",
        <div
          className="home__balance-buttons"
        >
          <button
            onClick={[Function]}
          >
            -
          </button>
          <button
            onClick={[Function]}
          >
            +
          </button>
        </div>,
      ],
      "className": "home",
    },
    "ref": null,
    "rendered": Array [
      "Your Balance: ",
      "£12",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              onClick={[Function]}
            >
              -
            </button>,
            <button
              onClick={[Function]}
            >
              +
            </button>,
          ],
          "className": "home__balance-buttons",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "-",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "-",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "+",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "+",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Your Balance: ",
          "£12",
          <div
            className="home__balance-buttons"
          >
            <button
              onClick={[Function]}
            >
              -
            </button>
            <button
              onClick={[Function]}
            >
              +
            </button>
          </div>,
        ],
        "className": "home",
      },
      "ref": null,
      "rendered": Array [
        "Your Balance: ",
        "£12",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                onClick={[Function]}
              >
                -
              </button>,
              <button
                onClick={[Function]}
              >
                +
              </button>,
            ],
            "className": "home__balance-buttons",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "-",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "-",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "+",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "+",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
